lib/rules/no-dupe-disjunctions.ts

```diff
@@ -9,6 +10,492 @@ import type {
+
+    const all = NFA.all({ maxCharacter: alternatives[0][0].maxCharacter })
+
+    for (let i = 1; i < alternatives.length; i++) {
+        const [nfa, alternative] = alternatives[i]
+
+        const overlapping = alternatives
+            .slice(0, i)
+            .filter(([otherNfa]) => !nfa.isDisjointWith(otherNfa))
+
+        if (overlapping.length >= 1) {
+            const othersNfa = unionAll(overlapping.map(([n]) => n))
+            const others = overlapping.map(([, a]) => a)
+
+            // Only checking for a subset relation is sufficient here.
+            // Duplicates are VERY unlikely. (Who would use alternatives
+            // like `a|a[^]*`?)
+
+            if (isSubsetOf(othersNfa, nfa)) {
+                yield { type: "PrefixSubset", alternative, others }
+            } else {
+                // The else case is interesting.
+                // It means that _some_ of the paths in the current alternative are useless.
+                // TODO: Decide whether this should be reported as well.
+            }
+        }
+
+        nfa.append(all)
+    }
+}
+
+/**
+ * Tries to find duplication in the given alternatives.
+ */
+function* findDuplicationNfa(
+    alternatives: Alternative[],
+    { hasNothingAfter, parser, ignoreOverlap }: Options,
+): Iterable<Result> {
+    const previous: [NFA, Alternative][] = []
+    for (let i = 0; i < alternatives.length; i++) {
+        const alternative = alternatives[i]
+        const nfa = toNFA(parser, alternative)
+
+        const overlapping = previous.filter(
+            ([otherNfa]) => !nfa.isDisjointWith(otherNfa),
+        )
+
+        if (overlapping.length >= 1) {
+            const othersNfa = unionAll(overlapping.map(([n]) => n))
+            const others = overlapping.map(([, a]) => a)
+
+            const relation = getSubsetRelation(nfa, othersNfa)
+            switch (relation) {
+                case SubsetRelation.leftEqualRight:
+                    if (others.length === 1) {
+                        // only return "duplicate" if there is only one other
+                        // alternative
+                        yield {
+                            type: "Duplicate",
+                            alternative,
+                            others: [others[0]],
+                        }
+                    } else {
+                        yield { type: "Subset", alternative, others }
+                    }
+                    break
+
+                case SubsetRelation.leftSubsetOfRight:
+                    yield { type: "Subset", alternative, others }
+                    break
+
+                case SubsetRelation.leftSupersetOfRight:
+                    yield { type: "Superset", alternative, others }
+                    break
+
+                case SubsetRelation.none:
+                case SubsetRelation.unknown:
+                    // TODO: we might want to report cases where the relation
```

Comment:
As I said in the [other comment](https://github.com/ota-meshi/eslint-plugin-regexp/pull/179/files#r619676707), the DFA construction might fail. If it fails, we don't know which message to display and currentls default to "overlap".

Should we display a special message if this happens?

The problem might be interesting for the user but I don't want to bother them with implementation details.

Reply:
I don't think DFA construction errors should report a report.

Reply:
Ok. I'll continue to report them as simple overlaps (if allowed by the user).